### PR TITLE
Add is_distinct to Aggregate (#296)

### PR DESCRIPTION
**Pull request by pjongy**
Enhanced Aggregate class for distinct option

## Description

For now, we can use Count function with:
Count('field', is_disitnct=True)
and then it might generate SQL like:
COUNT(DISTINC field)
by pypika:
Count().distinct()

## Motivation and Context

There were no solution for count with distinct about aggregated result

## How Has This Been Tested?

Execute count function with distinct flag in aggregating multiple backward relation situation.

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: tortoise/functions.py [hunk 1]
```diff
@@ -110,10 +113,32 @@ def resolve(self, model: "Type[Model]", table: Table) -> dict:
...
+    :param is_distinct: Flag for aggregate with distinction
     """
 
     database_func = AggregateFunction
 
+    def __init__(
+        self,
+        field: Union[str, F, ArithmeticExpression],
+        *default_values: Any,
+        is_distinct=False,
```
Comment (grigi) [inline tortoise/functions.py]: @lntuition Now here is a valid use-case to keep `Aggregate` separate from `Function`.
A function like `Length`, `Trim`, etc... doesn't make sense with a `distinct`, but aggregates really do.

Comment (grigi) [reply tortoise/functions.py]: @pjongy I feel an interface of `Count('somecol', distinct=True)` would read/write better.

**Comment by grigi** [comment]
Thank you for the PR @pjongy 
Other than the minor comment re `distinct` vs `is_distinct` the code looks good.
Could you please do a test? e.g. just confirm that the result changes correctly when distinct changes from `False` to `True`.
e.g. add multiple records, do an count() without and with a distinct, and check the result is correct (and different)

Also, please add a short feature description in the `CHANGELOG.rst` under section 0.16.0
and add yourself to the `CONTRIBUTORS.rst`

**Comment by pjongy** [comment]
First of all, thanks for your review

I accept your suggestion(distinct=True) is more reasonable. (FIxed)
I've add some test for COUNT(DISTINCT, column) (Added)

But while I'm working on, I found something going wrong in relation joining.
(Occurs same in CI: https://travis-ci.com/tortoise/tortoise-orm/jobs/295745089)

I think cause of test fails is wrong condition chain flows in while manuplate JOIN ~ ON condition
I expected (and it work in tortoise==0.15.8)
``` SQL
SELECT "post"."name","post"."id",COUNT(DISTINCT "comment"."id") "comments_count",COUNT(DISTINCT "favorite"."id") "favorites_count" FROM "post" LEFT OUTER JOIN "comment" ON "post"."id"="comment"."post_id" LEFT OUTER JOIN "favorite" ON "post"."id"="favorite"."post_id" WHERE "post"."id"=2 GROUP BY "post"."id" LIMIT 1
```
but real work ( == 0.15.9 upper)
``` SQL
SELECT "post"."name","post"."id",COUNT(DISTINCT "comment"."id") "comments_count",COUNT(DISTINCT "favorite"."id") "favorites_count" FROM "post" LEFT OUTER JOIN "comment" ON "post"."id"="comment"."post_id" LEFT OUTER JOIN "favorite" ON "comment"."id"="favorite"."post_id" WHERE "post"."id"=2 GROUP BY "post"."id" LIMIT 1
```
expect: LEFT OUTER JOIN "favorite" ON "post"."id"="favorite"
real: LEFT OUTER JOIN "favorite" ON "comment"."id"="favorite"

I think https://github.com/tortoise/tortoise-orm/commit/55618fa43cb18a9bd3db0a8959a40496a1c6f6f5 commit could affect in this situation
So, can you check this bug(?) for detail? @grigi 

Thank you.

**Comment by grigi** [comment]
I will have a look during the week. Thank you for finding where the regression happened, it will make it easier to fix the issue.

**Comment by lntuition** [comment]
@grigi @pjongy I found error point and fixed it. I'll share information after do my personal work.

**Comment by grigi** [comment]
Wow, that was fast. I was planning to look now.

**Comment by lntuition** [comment]
There was some wrong table assignment to `_resolve_annotate` method in `queryset`
With #316, I remove wrong table assignment and also checked it works well.
So just merge develop branch will remove this issue.

And also i have some reviews :) plz check it. Thank you for your hard work :smile:

File: CHANGELOG.rst [hunk 2]
```diff
@@ -98,6 +98,10 @@ New features:
 - Allow usage of ``F`` expressions to in annotations. (#301)
 - Now negative number with ``limit(...)`` and ``offset(...)`` raise ``ParamsError``. (#306)
 
+* Add distinct to Aggregate
+
+  Functions that inherits Aggregate can use with ``distinct`` flag
+
```
Comment (lntuition) [inline CHANGELOG.rst]: This CHANGELOG is also technically correct. but I think average user will not see the implication. 
Rather say `Add ability to supply ``distinct`` flag to Aggregate`? Just one line looks enough.

File: tests/testmodels.py [hunk 3]
```diff
@@ -637,3 +637,26 @@ class Principal(Model):
...
+    post: fields.ForeignKeyRelation[Post] = fields.ForeignKeyField(
+        "models.Post", related_name="comments"
+    )
+
+
+class Favorite(Model):
+    id = fields.IntField(pk=True)
+    post: fields.ForeignKeyRelation[Post] = fields.ForeignKeyField(
+        "models.Post", related_name="favorites"
+    )
```
Comment (lntuition) [inline tests/testmodels.py]: I think adding new models is not essential. Also Current test model is enough huge. How about reusing `DoubleFK` model instead?

File: tests/test_aggregation_with_distinct.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+        school_with_count = (
+            await Post.filter(id=post.id)
+            .annotate(comments_count=Count("comments"), favorites_count=Count("favorites"))
+            .first()
+        )
+
+        self.assertEqual(school_with_count.comments_count, 6)
+        self.assertEqual(school_with_count.favorites_count, 6)
+        # comments_count and favorites_count will same if no distinct
```
Comment (lntuition) [inline tests/test_aggregation_with_distinct.py]: We do not need to test this here. It is part of aggregation not aggregation with distinct. How about removing this? And also I think adding new TestClass is not essential. How about moving this testsuite to `test_aggregation.py` (inside of `TestAggregation`) ?

File: CHANGELOG.rst (hunk 2 above)
Comment (pjongy) [reply CHANGELOG.rst]: Done.
Thx for your suggestion

File: tests/testmodels.py (hunk 3 above)
Comment (pjongy) [reply tests/testmodels.py]: I think models we need are "multiple backward relation" not "multiple foriegn key relation" that is your suggestion. So, how about using just `Tournament` that has two backward relation?
(And your model redundant model removing opinion is acceptable. - Done)

File: tests/test_aggregation_with_distinct.py (hunk 4 above)
Comment (pjongy) [reply tests/test_aggregation_with_distinct.py]: Moving this test suite to test_aggregation is reasonable. - Done.
And my idea for asserting without `distinct` was to show comparison opposites `distinct`, but I think it does not feat for test as your opinion. - Done.

File: tests/testmodels.py (hunk 3 above)
Comment (lntuition) [reply tests/testmodels.py]: I misunderstand point. Anywhere change looks good :+1:

**Pull request merged**